src/replication.c

```diff
@@ -2386,6 +2481,353 @@ char *sendCommandArgv(connection *conn, int argc, char **argv, size_t *argv_lens
+    serverAssert(conn == server.repl_transfer_s);
+    /* Init the main connection for psync */
+    conn->state = CONN_STATE_CONNECTED;
+    server.repl_state = REPL_SEC_CONN_SEND_PSYNC;
+    serverAssert(connSetReadHandler(server.psync_master->conn, syncWithMaster) != C_ERR);
+    syncWithMaster(conn);
+    return C_OK;
+}
+
+int prepareRdbConnectionForRdbLoad(connection *conn) {
+    serverAssert(isRdbConnectionSync(conn));
+    /* Set connection handler for RDB */
+    serverAssert(connSetReadHandler(conn, readSyncBulkPayload) != C_ERR);
+    server.repl_transfer_size = -1;
+    server.repl_transfer_read = 0;
+    server.repl_transfer_last_fsync_off = 0;
+    server.repl_transfer_lastio = server.unixtime;
+    return C_OK;
+}
+
+int processEndOffsetResponse(char* response) {
+    long long reploffset;
+    char master_replid[CONFIG_RUN_ID_SIZE+1];
+    int dbid;
+    char endoff_format[30];
+    snprintf(endoff_format, sizeof(endoff_format), "$ENDOFF:%%lld %%%ds %%d", CONFIG_RUN_ID_SIZE);
+    if (sscanf(response, endoff_format, &reploffset, master_replid, &dbid) != 3) {
+        return C_ERR;
+    }
+    server.master_initial_offset = reploffset;
+    server.psync_master = replicationCreateMasterClient(server.repl_transfer_s, dbid);
+    memcpy(server.psync_master->replid, master_replid, CONFIG_RUN_ID_SIZE);
+    server.psync_master->reploff = reploffset;
+    server.psync_master->read_reploff = reploffset;
+    return C_OK;
+}
+
+/* This connection handler is used to initialize the RDB connection (repl-rdb-channel sync). 
+ * Once a replica with repl-rdb-channel enabled, denied from PSYNC with its primary, 
+ * fullSyncWithMaster begins its role. The connection handler prepare server.repl_full_sync_s
+ * for a rdb stream, and server.repl_transfer_s for replication data stream. */
+void fullSyncWithMaster(connection* conn) {
+    char *err = NULL;
+    serverAssert(conn == server.repl_full_sync_s);
+
+    /* If this event fired after the user turned the instance into a master
+     * with SLAVEOF NO ONE we must just return ASAP. */
+    if (server.repl_state == REPL_STATE_NONE) {
+        goto error;
+    }
+
+    /* Check for errors in the socket: after a non blocking connect() we
+     * may find that the socket is in error state. */
+    if (connGetState(conn) != CONN_STATE_CONNECTED) {
+        serverLog(LL_WARNING,"Error condition on socket for RDB-CHANNEL-SYNC: %s",
+                connGetLastError(conn));
+        goto error;
+    }
+
+    /* Send replica capabilities */
+    if (server.repl_state == REPL_STATE_RECEIVE_PSYNC_REPLY) {
+        serverLog(LL_DEBUG, "Received first reply from primary using rdb connection. Sending capa");
+        
+        /* Send replica lisening port to master for clarification */
+        sds portstr = getReplicaPortString();        
+        err = sendCommand(conn, "REPLCONF", "capa", "eof",
+                            "rdb-only", "1", "end-offset", "1", "listening-port", portstr, NULL);
```

Comment:
> if i understand the current code, we do a PSYNC (attempting a partial sync), then if it results in full-sync we abort it, and start over with a 2 channel sync even if the server doesn't support 2 channel sync.

Correct, however we have already verified remote capability in some way. When we try to psync the first time, we call `REPLCONF no-fullsync`, so at this point we already know that master supports rdb-channel. This command can't fail because of old master version. 

Perhaps we should remove the verification from REPL_RDB_CONN_RECEIVE_REPLCONF_REPLY (15 lines below). I believe that, although this scenario is impossible, it keeps the implementation complete.

Reply:
The fallback mechanism is implemented in this revision. 
It happens when the replica receives `REPLCONF PSYNC-ONLY` response (will change it to replconf main-channel).

I will make it clearer by verifying the default value of the master response (`+OK`) in this way you can't miss it :).

Reply:
still, please invest some time in editing the top comment to describe both the design and mechanics of the new mechanism, as well as the fallback.
it'll be easier to make sense of the code after reading the description.

Reply:
the chart in the top comment now handles it, right? so we can dismiss this thread?
however, the image you posted there is a bit hard to read (due to scaling), and when i click on it to get the larger version, i get error 404.

on the other hand, we don't want the code base (and future generations) to rely looking at that image in the PR, we need to convert that diagram to ASCII art and also include it in the code (unless you already did).

Reply:
I fixed the diagram. Regarding top comment for high level overview, I wanted to put this explanation somewhere in the code (not sure yet where does it belong).
```
/* 
 * RDB-Channel sync high level interface change:
 * RDB-Channel sync begins when the replica sends a REPLCONF MAINCONN  to the primary during initial 
 * handshake. This allows the replica to verify whether the master supports rdb-channel sync and, if 
 * so, state that this is the replica's main connection, which is not used for snapshots. 
 * When replica lacks sufficient data for PSYNC, the master will send -FULLSYNCNEEDED response instead 
 * of RDB. As a next step, the replica creates a new connection (rdb-channel) and configures it against 
 * the master with the appropriate capabilities and requirements. The replica then requests a sync 
 * using the RDB connection. Prior to forking, the master sends the replica the snapshot's end offset, 
 * and attaches the replica to the replication backlog to keep repl data until the replica requests 
 * psync. The replica uses the main connection to request a PSYNC starting at the snapshot end offset. 
 * The master main threads sends incremental changes via the main connection, while the bgsave process 
 * sends the RDB directly to the replica via the rdb-connection. As for the replica, the incremental 
 * changes are stored on a local buffer, while the RDB is loaded into memory. Once the replica 
 * completes loading the rdb, it drops the rdb-connection and streams the accumulated incremental 
 * changes into memory. Repl steady state continues normally.
 */

```
Its a good idea to add the diagram as well but it isn't enough to cover all interface changes.